diff --git a/api/moz.build b/api/moz.build
new file mode 100644
index 0000000000000..d9ee60c924c39
--- /dev/null
+++ b/api/moz.build
@@ -0,0 +1,87 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+EXPORTS.opentelemetry += [
+    "include/opentelemetry/config.h",
+    "include/opentelemetry/version.h",
+]
+
+EXPORTS.opentelemetry.common += [
+    "include/opentelemetry/common/attribute_value.h",
+    "include/opentelemetry/common/key_value_iterable.h",
+    "include/opentelemetry/common/key_value_iterable_view.h",
+    "include/opentelemetry/common/kv_properties.h",
+    "include/opentelemetry/common/macros.h",
+    "include/opentelemetry/common/spin_lock_mutex.h",
+    "include/opentelemetry/common/string_util.h",
+    "include/opentelemetry/common/timestamp.h",
+]
+
+EXPORTS.opentelemetry.context += [
+    "include/opentelemetry/context/context.h",
+    "include/opentelemetry/context/context_value.h",
+    "include/opentelemetry/context/runtime_context.h",
+]
+
+EXPORTS.opentelemetry.detail += [
+    "include/opentelemetry/detail/preprocessor.h",
+]
+
+# EXPORTS.opentelemetry.logs += [
+#   "include/opentelemetry/logs/event_id.h",
+#   "include/opentelemetry/logs/event_logger.h",
+#   "include/opentelemetry/logs/event_logger_provider.h",
+#   "include/opentelemetry/logs/log_record.h",
+#   "include/opentelemetry/logs/logger.h",
+#   "include/opentelemetry/logs/logger_provider.h",
+#   "include/opentelemetry/logs/logger_type_traits.h",
+#   "include/opentelemetry/logs/noop.h",
+#   "include/opentelemetry/logs/provider.h",
+#   "include/opentelemetry/logs/severity.h",
+# ]
+
+# Probably this is where we can bind to custom mozilla types.
+EXPORTS.opentelemetry.nostd += [
+    "include/opentelemetry/nostd/function_ref.h",
+    "include/opentelemetry/nostd/shared_ptr.h",
+    "include/opentelemetry/nostd/span.h",
+    "include/opentelemetry/nostd/string_view.h",
+    "include/opentelemetry/nostd/type_traits.h",
+    "include/opentelemetry/nostd/unique_ptr.h",
+    "include/opentelemetry/nostd/utility.h",
+    "include/opentelemetry/nostd/variant.h",
+]
+
+EXPORTS.opentelemetry.std = [
+    "include/opentelemetry/std/shared_ptr.h",
+    "include/opentelemetry/std/span.h",
+    "include/opentelemetry/std/string_view.h",
+    "include/opentelemetry/std/type_traits.h",
+    "include/opentelemetry/std/unique_ptr.h",
+    "include/opentelemetry/std/utility.h",
+    "include/opentelemetry/std/variant.h",
+]
+
+EXPORTS.opentelemetry.trace += [
+    "include/opentelemetry/trace/context.h",
+    "include/opentelemetry/trace/default_span.h",
+    "include/opentelemetry/trace/noop.h",
+    "include/opentelemetry/trace/provider.h",
+    "include/opentelemetry/trace/scope.h",
+    "include/opentelemetry/trace/semantic_conventions.h",
+    "include/opentelemetry/trace/span.h",
+    "include/opentelemetry/trace/span_context.h",
+    "include/opentelemetry/trace/span_context_kv_iterable.h",
+    "include/opentelemetry/trace/span_context_kv_iterable_view.h",
+    "include/opentelemetry/trace/span_id.h",
+    "include/opentelemetry/trace/span_metadata.h",
+    "include/opentelemetry/trace/span_startoptions.h",
+    "include/opentelemetry/trace/trace_flags.h",
+    "include/opentelemetry/trace/trace_id.h",
+    "include/opentelemetry/trace/trace_state.h",
+    "include/opentelemetry/trace/tracer.h",
+    "include/opentelemetry/trace/tracer_provider.h",
+]
diff --git a/exporters/moz.build b/exporters/moz.build
new file mode 100644
index 0000000000000..07ff09c58e1bc
--- /dev/null
+++ b/exporters/moz.build
@@ -0,0 +1,9 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DIRS += [
+  "ostream",
+]
diff --git a/exporters/ostream/moz.build b/exporters/ostream/moz.build
new file mode 100644
index 0000000000000..b0bf91d243cc0
--- /dev/null
+++ b/exporters/ostream/moz.build
@@ -0,0 +1,32 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("opentelemetry_exporter_ostream_span")
+
+SOURCES += [
+    "src/span_exporter.cc",
+    "src/span_exporter_factory.cc",
+]
+
+LOCAL_INCLUDES += [
+    "/third_party/opentelemetry-cpp/api/include",
+    "/third_party/opentelemetry-cpp/exporters/ostream/include",
+    "/third_party/opentelemetry-cpp/sdk/include",
+]
+
+USE_LIBS += [
+    "gkopentelemetry",
+]
+
+EXPORTS.opentelemetry.exporters.ostream += [
+    "include/opentelemetry/exporters/ostream/common_utils.h",
+    "include/opentelemetry/exporters/ostream/log_record_exporter.h",
+    "include/opentelemetry/exporters/ostream/log_record_exporter_factory.h",
+    "include/opentelemetry/exporters/ostream/metric_exporter.h",
+    "include/opentelemetry/exporters/ostream/metric_exporter_factory.h",
+    "include/opentelemetry/exporters/ostream/span_exporter.h",
+    "include/opentelemetry/exporters/ostream/span_exporter_factory.h",
+]
diff --git a/moz.build b/moz.build
new file mode 100644
index 0000000000000..092d0ad4421a4
--- /dev/null
+++ b/moz.build
@@ -0,0 +1,11 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DIRS += [
+    "api",
+    "exporters",
+    "sdk",
+]
diff --git a/moz.yaml b/moz.yaml
index 363dbc2eb4d10..c84ab6d486b4f 100644
--- a/moz.yaml
+++ b/moz.yaml
@@ -18,6 +18,9 @@ vendoring:
     url: https://github.com/open-telemetry/opentelemetry-cpp
     source-hosting: github
 
+    patches:
+      - patches/0001-Instrument-lib-using-mozilla-build-system.patch
+
     exclude:
       - "**"
 
diff --git a/sdk/moz.build b/sdk/moz.build
new file mode 100644
index 0000000000000..0133a3857b800
--- /dev/null
+++ b/sdk/moz.build
@@ -0,0 +1,93 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+DIRS += ["src"]
+
+EXPORTS.opentelemetry.sdk.common += [
+    "include/opentelemetry/sdk/common/atomic_shared_ptr.h",
+    "include/opentelemetry/sdk/common/atomic_unique_ptr.h",
+    "include/opentelemetry/sdk/common/attribute_utils.h",
+    "include/opentelemetry/sdk/common/attributemap_hash.h",
+    "include/opentelemetry/sdk/common/base64.h",
+    "include/opentelemetry/sdk/common/circular_buffer.h",
+    "include/opentelemetry/sdk/common/circular_buffer_range.h",
+    "include/opentelemetry/sdk/common/empty_attributes.h",
+    "include/opentelemetry/sdk/common/env_variables.h",
+    "include/opentelemetry/sdk/common/exporter_utils.h",
+    "include/opentelemetry/sdk/common/global_log_handler.h",
+]
+
+EXPORTS.opentelemetry.sdk.instrumentationscope += [
+    "include/opentelemetry/sdk/instrumentationscope/instrumentation_scope.h",
+]
+
+# EXPORTS.opentelemetry.sdk.logs += [
+#     "include/opentelemetry/sdk/logs/batch_log_record_processor.h",
+#     "include/opentelemetry/sdk/logs/batch_log_record_processor_factory.h",
+#     "include/opentelemetry/sdk/logs/batch_log_record_processor_options.h",
+#     "include/opentelemetry/sdk/logs/event_logger.h",
+#     "include/opentelemetry/sdk/logs/event_logger_provider.h",
+#     "include/opentelemetry/sdk/logs/event_logger_provider_factory.h",
+#     "include/opentelemetry/sdk/logs/exporter.h",
+#     "include/opentelemetry/sdk/logs/logger.h",
+#     "include/opentelemetry/sdk/logs/logger_context.h",
+#     "include/opentelemetry/sdk/logs/logger_context_factory.h",
+#     "include/opentelemetry/sdk/logs/logger_provider.h",
+#     "include/opentelemetry/sdk/logs/logger_provider_factory.h",
+#     "include/opentelemetry/sdk/logs/multi_log_record_processor.h",
+#     "include/opentelemetry/sdk/logs/multi_log_record_processor_factory.h",
+#     "include/opentelemetry/sdk/logs/multi_recordable.h",
+#     "include/opentelemetry/sdk/logs/processor.h",
+#     "include/opentelemetry/sdk/logs/read_write_log_record.h",
+#     "include/opentelemetry/sdk/logs/readable_log_record.h",
+#     "include/opentelemetry/sdk/logs/recordable.h",
+#     "include/opentelemetry/sdk/logs/simple_log_record_processor.h",
+#     "include/opentelemetry/sdk/logs/simple_log_record_processor_factory.h",
+# ]
+
+EXPORTS.opentelemetry.sdk.resource += [
+    "include/opentelemetry/sdk/resource/resource.h",
+    "include/opentelemetry/sdk/resource/resource_detector.h",
+    "include/opentelemetry/sdk/resource/semantic_conventions.h",
+]
+
+EXPORTS.opentelemetry.sdk.trace += [
+    "include/opentelemetry/sdk/trace/batch_span_processor.h",
+    "include/opentelemetry/sdk/trace/batch_span_processor_factory.h",
+    "include/opentelemetry/sdk/trace/batch_span_processor_options.h",
+    "include/opentelemetry/sdk/trace/exporter.h",
+    "include/opentelemetry/sdk/trace/id_generator.h",
+    "include/opentelemetry/sdk/trace/multi_recordable.h",
+    "include/opentelemetry/sdk/trace/multi_span_processor.h",
+    "include/opentelemetry/sdk/trace/processor.h",
+    "include/opentelemetry/sdk/trace/random_id_generator.h",
+    "include/opentelemetry/sdk/trace/random_id_generator_factory.h",
+    "include/opentelemetry/sdk/trace/recordable.h",
+    "include/opentelemetry/sdk/trace/sampler.h",
+    "include/opentelemetry/sdk/trace/simple_processor.h",
+    "include/opentelemetry/sdk/trace/simple_processor_factory.h",
+    "include/opentelemetry/sdk/trace/span_data.h",
+    "include/opentelemetry/sdk/trace/tracer.h",
+    "include/opentelemetry/sdk/trace/tracer_context.h",
+    "include/opentelemetry/sdk/trace/tracer_context_factory.h",
+    "include/opentelemetry/sdk/trace/tracer_provider.h",
+    "include/opentelemetry/sdk/trace/tracer_provider_factory.h",
+]
+
+EXPORTS.opentelemetry.sdk.trace.samplers += [
+    "include/opentelemetry/sdk/trace/samplers/always_off.h",
+    "include/opentelemetry/sdk/trace/samplers/always_off_factory.h",
+    "include/opentelemetry/sdk/trace/samplers/always_on.h",
+    "include/opentelemetry/sdk/trace/samplers/always_on_factory.h",
+    "include/opentelemetry/sdk/trace/samplers/parent.h",
+    "include/opentelemetry/sdk/trace/samplers/parent_factory.h",
+    "include/opentelemetry/sdk/trace/samplers/trace_id_ratio.h",
+    "include/opentelemetry/sdk/trace/samplers/trace_id_ratio_factory.h",
+]
+
+EXPORTS.opentelemetry.sdk.version += ["include/opentelemetry/sdk/version/version.h"]
+
+FINAL_LIBRARY = "xul"
diff --git a/sdk/src/common/moz.build b/sdk/src/common/moz.build
new file mode 100644
index 0000000000000..5e927019daf71
--- /dev/null
+++ b/sdk/src/common/moz.build
@@ -0,0 +1,27 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("opentelemetry_common")
+
+LOCAL_INCLUDES += [
+  "/third_party/opentelemetry-cpp/api/include",
+  "/third_party/opentelemetry-cpp/sdk",
+  "/third_party/opentelemetry-cpp/sdk/include",
+]
+
+SOURCES += [
+    "base64.cc",
+    "env_variables.cc",
+    "global_log_handler.cc",
+    "random.cc",
+]
+
+if CONFIG["OS_ARCH"] == "WINNT":
+  SOURCES += [ "platform/fork_windows.cc" ]
+else:
+  SOURCES += [ "platform/fork_unix.cc" ]
+
+FINAL_LIBRARY = "gkopentelemetry"
diff --git a/sdk/src/moz.build b/sdk/src/moz.build
new file mode 100644
index 0000000000000..8be8fefa467f2
--- /dev/null
+++ b/sdk/src/moz.build
@@ -0,0 +1,25 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("gkopentelemetry")
+
+DIRS += [
+    "common",
+    # "logs",
+    "resource",
+    "trace",
+    "version",
+]
+
+USE_LIBS += [
+    "opentelemetry_common",
+    # "opentelemetry_logs",
+    "opentelemetry_resource",
+    "opentelemetry_trace",
+    "opentelemetry_version",
+]
+
+FINAL_LIBRARY = "xul"
diff --git a/sdk/src/resource/moz.build b/sdk/src/resource/moz.build
new file mode 100644
index 0000000000000..e8b394070caca
--- /dev/null
+++ b/sdk/src/resource/moz.build
@@ -0,0 +1,19 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("opentelemetry_resource")
+
+LOCAL_INCLUDES += [
+  "/third_party/opentelemetry-cpp/api/include",
+  "/third_party/opentelemetry-cpp/sdk/include",
+]
+
+SOURCES += [
+  "resource.cc",
+  "resource_detector.cc",
+]
+
+FINAL_LIBRARY = "gkopentelemetry"
\ No newline at end of file
diff --git a/sdk/src/trace/moz.build b/sdk/src/trace/moz.build
new file mode 100644
index 0000000000000..853b0746fff5d
--- /dev/null
+++ b/sdk/src/trace/moz.build
@@ -0,0 +1,36 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("opentelemetry_trace")
+
+LOCAL_INCLUDES += [
+  "/third_party/opentelemetry-cpp/api/include",
+  "/third_party/opentelemetry-cpp/sdk",
+  "/third_party/opentelemetry-cpp/sdk/include",
+]
+
+SOURCES += [
+  "batch_span_processor.cc",
+  "batch_span_processor_factory.cc",
+  "exporter.cc",
+  "random_id_generator.cc",
+  "random_id_generator_factory.cc",
+  "samplers/always_off_factory.cc",
+  "samplers/always_on_factory.cc",
+  "samplers/parent.cc",
+  "samplers/parent_factory.cc",
+  "samplers/trace_id_ratio.cc",
+  "samplers/trace_id_ratio_factory.cc",
+  "simple_processor_factory.cc",
+  "span.cc",
+  "tracer.cc",
+  "tracer_context.cc",
+  "tracer_context_factory.cc",
+  "tracer_provider.cc",
+  "tracer_provider_factory.cc",
+]
+
+FINAL_LIBRARY = "gkopentelemetry"
diff --git a/sdk/src/version/moz.build b/sdk/src/version/moz.build
new file mode 100644
index 0000000000000..48eee1250aa3f
--- /dev/null
+++ b/sdk/src/version/moz.build
@@ -0,0 +1,18 @@
+# -*- Mode: python; indent-tabs-mode: nil; tab-width: 40 -*-
+# vim: set filetype=python:
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+Library("opentelemetry_version")
+
+LOCAL_INCLUDES += [
+  "/third_party/opentelemetry-cpp/api/include",
+  "/third_party/opentelemetry-cpp/sdk/include",
+]
+
+SOURCES += [
+  "version.cc",
+]
+
+FINAL_LIBRARY = "gkopentelemetry"
\ No newline at end of file
